### PR TITLE
feat(Boards,CMSIS,Examples,PeriphDrivers): Initial commit for MAX32657

### PR DESCRIPTION
**Pull request by sihyung-maxim**
### Description

This is the initial commit for the MAX32657.

The MAX32657 (ME30) SDK files are still in development and not for official, public use.

Please do not delete the branch after merging this PR.

### Checklist Before Requesting Review

- [ ] PR Title follows correct guidelines.
- [ ] Description of changes and all other relevant information.
- [ ] (Optional) Link any related GitHub issues [using a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [ ] (Optional) Provide info on any relevant functional testing/validation.  For API changes or significant features, this is not optional.

### CODE REVIEW

File: Libraries/CMSIS/Device/Maxim/MAX32657/Include/max32657.h [hunk 1]
```diff
@@ -0,0 +1,907 @@
...
+#define __MPU_PRESENT 1U /**< Presence of MPU  */
+#define __SAUREGION_PRESENT 1U /**< Presence of FPU  */
+#define __TZ_PRESENT 1U /**< Presence of TrustZone */
+#define __VTOR_PRESENT 1U /**< Presence of VTOR register in SCB  */
+#define __NVIC_PRIO_BITS 4U /**< NVIC interrupt priority bits */
+#define __Vendor_SysTickConfig 0U /**< Is 1 if different SysTick counter is used */
+
+#include <core_cm33.h>
+#include <arm_cmse.h>
+#define IS_SECURE_ENVIRONMENT (defined(__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U))
```
Comment (ozersa) [inline Libraries/CMSIS/Device/Maxim/MAX32657/Include/max32657.h:156]: This definition causes issues on zephyr side.
image

**Comment by ozersa** [comment]
This PR shall be merged after below one, to this changes being pushed on hal_adi too.
https://github.com/analogdevicesinc/msdk/pull/994

File: Libraries/CMSIS/Device/Maxim/MAX32657/Include/max32657.h (hunk 1 above)
Comment (Jake-Carter) [reply Libraries/CMSIS/Device/Maxim/MAX32657/Include/max32657.h]: @ozersa fixed in https://github.com/analogdevicesinc/msdk/pull/1012/commits/f040b6dea9e524ea9de0506d31fc2328bbf758a2

**Comment by Jake-Carter** [comment]
@sihyung-maxim there are some DMA-related build errors if you attempt to build with `MSECURITY_MODE = NONSECURE`, can you take a look?

File: Libraries/PeriphDrivers/Source/UART/uart_revb.h [hunk 2]
```diff
@@ -74,14 +74,15 @@ int MXC_UART_RevB_Read(mxc_uart_revb_regs_t *uart, uint8_t *buffer, int *len);
...
+int MXC_UART_RevB_ReadRXFIFODMA(mxc_uart_revb_regs_t *uart, mxc_dma_regs_t *dma,
+                                unsigned char *bytes, unsigned int len,
                                 mxc_uart_dma_complete_cb_t callback, mxc_dma_config_t config);
 unsigned int MXC_UART_RevB_GetRXFIFOAvailable(mxc_uart_revb_regs_t *uart);
 unsigned int MXC_UART_RevB_WriteTXFIFO(mxc_uart_revb_regs_t *uart, const unsigned char *bytes,
                                        unsigned int len);
-int MXC_UART_RevB_WriteTXFIFODMA(mxc_uart_revb_regs_t *uart, const unsigned char *bytes,
-                                 unsigned int len, mxc_uart_dma_complete_cb_t callback,
-                                 mxc_dma_config_t config);
+int MXC_UART_RevB_WriteTXFIFODMA(mxc_uart_revb_regs_t *uart, mxc_dma_regs_t *dma,
```
Comment (rcornall-adi) [inline Libraries/PeriphDrivers/Source/UART/uart_revb.h:83]: I think you might need to add *dma parameter also to `Libraries/PeriphDrivers/Include/MAX32655/uart.h` and other ones that that use revb uart?

I'm getting an error trying to build the zephyr samples :

```
/home/rcornall/zephyrproject/modules/hal/adi/MAX/Libraries/PeriphDrivers/Source/UART/uart_revb.c:1084:13: error: too many arguments to function 'MXC_UART_ReadRXFIFODMA'
 1084 |         if (MXC_UART_ReadRXFIFODMA((mxc_uart_regs_t *)(req->uart), dma, req->rxData, req->rxLen, NULL) !=
      |             ^~~~~~~~~~~~~~~~~~~~~~
```

**Comment by sihyung-maxim** [comment]
> @sihyung-maxim there are some DMA-related build errors if you attempt to build with `MSECURITY_MODE = NONSECURE`, can you take a look?

Dang, I really wanted to avoid using `TARGET_NUM == 32657` and use `MXC_DMA_INSTANCES > 1` for the UART DMA functions. The DMA function prototypes are different when one DMA instance is available `int MXC_DMA_Init(void)` vs more than one DMA instance available `MXC_DMA_Init(mxc_dma_regs_t *dma)`. But for the ME30 non-secure build, although only one DMA instance is available, it still uses the multiple-DMA type functions: `MXC_DMA_Init(mxc_dma_regs_t *dma)`...

I reverted back the conditional directives. :(

File: Libraries/PeriphDrivers/Source/UART/uart_revb.h (hunk 2 above)
Comment (sihyung-maxim) [reply Libraries/PeriphDrivers/Source/UART/uart_revb.h:83]: Good to hear!

File: Libraries/CMSIS/Device/Maxim/MAX32657/Include/uart_regs.h [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+#define MXC_F_UART_INTEN_RX_OV                         ((uint32_t)(0x1UL << MXC_F_UART_INTEN_RX_OV_POS)) /**< INTEN_RX_OV Mask */
+
+#define MXC_F_UART_INTEN_RX_THD_POS                    4 /**< INTEN_RX_THD Position */
+#define MXC_F_UART_INTEN_RX_THD                        ((uint32_t)(0x1UL << MXC_F_UART_INTEN_RX_THD_POS)) /**< INTEN_RX_THD Mask */
+
+#define MXC_F_UART_INTEN_TX_OB_POS                     5 /**< INTEN_TX_OB Position */
+#define MXC_F_UART_INTEN_TX_OB                         ((uint32_t)(0x1UL << MXC_F_UART_INTEN_TX_OB_POS)) /**< INTEN_TX_OB Mask */
+
+#define MXC_F_UART_INTEN_TX_THD_POS                    6 /**< INTEN_TX_THD Position */
+#define MXC_F_UART_INTEN_TX_THD                        ((uint32_t)(0x1UL << MXC_F_UART_INTEN_TX_THD_POS)) /**< INTEN_TX_THD Mask */
```
Comment (rcornall-adi) [inline Libraries/CMSIS/Device/Maxim/MAX32657/Include/uart_regs.h:237]: I think this should still be `INTEN_TX_HE` as on the max32655?

image

**Review by lorne-maxim — APPROVED** [review]

**Pull request merged**